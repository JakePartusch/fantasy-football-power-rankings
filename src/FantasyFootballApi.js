import axios from 'axios';
import axiosCookieJarSupport from '@3846masa/axios-cookiejar-support';
import tough from 'tough-cookie';

export class FantasyFootballApi {
    constructor() {
        axiosCookieJarSupport(axios);
        const cookieJar = new tough.CookieJar();

        axios.defaults.jar = cookieJar;
        axios.defaults.withCredentials = true;
    }

    getPowerRankings = async (leagueId, seasonId) => {
        const weeklyScoreDataForSeason = await this.getWeeklyScoreDataForSeason(leagueId, seasonId);
        const weeklyWinsForSeason = this.calculateWeeklyWinsForSeason(weeklyScoreDataForSeason);
        return this.calculateSeasonWinTotal(weeklyWinsForSeason);
    }

    getUserData = async (leagueId, seasonId) => {
        var response = await axios.get(`http://games.espn.com/ffl/api/v2/leagueSettings?leagueId=${leagueId}&seasonId=${seasonId}`);
        return Object.values(response.data.leaguesettings.teams).map(team => (
            {
                id: team.teamId,
                logoUrl: team.logoUrl ? team.logoUrl: "http://www.freeiconspng.com/uploads/clipart--person-icon--cliparts-15.png",
                owner: `${team.owners[0].firstName} ${team.owners[0].lastName}`,
                name: `${team.teamLocation} ${team.teamNickname}`
            }
        ))
    }

    calculateSeasonWinTotal = (weeklyWinsForSeason) => {
        let seasonTotal = [];
        weeklyWinsForSeason.forEach(weekWins => {
            weekWins.forEach(team => {
                if(seasonTotal.find(seasonTotalTeam => seasonTotalTeam.id === team.id)) {
                    let element = seasonTotal.find(seasonTotalTeam => seasonTotalTeam.id === team.id);
                    seasonTotal[seasonTotal.indexOf(element)].wins += team.wins;
                    seasonTotal[seasonTotal.indexOf(element)].losses += team.losses  
                } else {
                    seasonTotal.push(team)
                }
            })
        });
        seasonTotal.sort((a, b) => { 
            return b.wins - a.wins;
        });
        return seasonTotal;
    }

    //Given the weekly score data is already sorted by score from low -> high
    calculateWeeklyWinsForSeason = (weeklyScoreDataForSeason) => {
        return weeklyScoreDataForSeason
                .map(singleWeekScoreData => singleWeekScoreData
                    .map((teamScoreData, index) => ({
                        wins: index,
                        losses: singleWeekScoreData.length - 1 - index,
                        id: teamScoreData.id,
                        name: teamScoreData.name
                    })))
    }

    getWeeklyScoreDataForSeason = async (leagueId, seasonId) => {
        let seasonData = [];
        var response = await axios.get(`http://games.espn.com/ffl/api/v2/leagueSettings?leagueId=${leagueId}&seasonId=${seasonId}`);
        const weeksInSeason = response.data.leaguesettings.regularSeasonMatchupPeriodCount;
        for (let i = 1; i <= weeksInSeason; i++) {
            const weekScores = await this.getWeekScores(leagueId, seasonId, i)
            weekScores.sort((a, b) => { 
                return a.score - b.score;
            });
            seasonData.push(weekScores);
        }
        return seasonData;
    }

    getWeekScores = async (leagueId, seasonId, week) => {
        var response = await axios.get(`http://games.espn.com/ffl/api/v2/scoreboard?leagueId=${leagueId}&seasonId=${seasonId}&matchupPeriodId=${week}`);
        let matchups = response.data.scoreboard.matchups;
        return matchups
            .filter(matchup => matchup.winner !== 'undecided' || matchup.bye )
            .reduce((acc, matchup) => acc.concat(matchup.teams), [])
            .map(team => ({
                id: team.teamId, 
                score: team.score,
                name: `${team.team.teamLocation} ${team.team.teamNickname}`
            }))
            .filter(team => team.score !== 0);
    }
}

