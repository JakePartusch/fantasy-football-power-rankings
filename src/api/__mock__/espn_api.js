const fs = require('fs');

export const scoreboard_response = week => {
  return require('./scoreboard-' + week + '.json');
};
