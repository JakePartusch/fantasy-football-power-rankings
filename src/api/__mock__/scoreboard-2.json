{
  "metadata": {
    "serverDate": "2018-10-18T01:37:07.597Z",
    "defaults": {},
    "seasonId": "2018",
    "leagueId": "452354",
    "dateModifiedLeague": "2018-10-18T00:44:23.823Z",
    "status": "regular-season"
  },
  "scoreboard": {
    "matchups": [
      {
        "winner": "undecided",
        "teams": [
          {
            "score": 0,
            "teamId": 6,
            "team": {
              "division": { "size": 13, "divisionName": "East", "divisionId": 0 },
              "teamAbbrev": "JUBE",
              "waiverRank": 3,
              "teamId": 6,
              "record": {
                "pointsAgainst": 588.9,
                "divisionTies": 0,
                "streakLength": 2,
                "homeTies": 0,
                "overallWins": 2,
                "pointsFor": 517.46,
                "divisionPercentage": 0.33333,
                "homeLosses": 2,
                "awayWins": 0,
                "overallPercentage": 0.33333,
                "overallLosses": 4,
                "divisionLosses": 4,
                "divisionWins": 2,
                "homeWins": 2,
                "awayPercentage": 0,
                "streakType": 2,
                "awayLosses": 2,
                "overallTies": 0,
                "divisionStanding": 11,
                "overallStanding": 11,
                "homePercentage": 0,
                "awayTies": 0
              },
              "teamLocation": "Team",
              "teamNickname": "Juberian"
            },
            "playerIDs": [],
            "home": true
          }
        ],
        "bye": true
      },
      {
        "winner": "undecided",
        "teams": [
          {
            "score": 0,
            "teamId": 4,
            "team": {
              "division": { "size": 13, "divisionName": "East", "divisionId": 0 },
              "teamAbbrev": "SCHL",
              "waiverRank": 9,
              "teamId": 4,
              "record": {
                "pointsAgainst": 610.76,
                "divisionTies": 0,
                "streakLength": 1,
                "homeTies": 0,
                "overallWins": 4,
                "pointsFor": 635.94,
                "divisionPercentage": 0.66667,
                "homeLosses": 2,
                "awayWins": 3,
                "overallPercentage": 0.66667,
                "overallLosses": 2,
                "divisionLosses": 2,
                "divisionWins": 4,
                "homeWins": 1,
                "awayPercentage": 0,
                "streakType": 2,
                "awayLosses": 0,
                "overallTies": 0,
                "divisionStanding": 5,
                "overallStanding": 5,
                "homePercentage": 0,
                "awayTies": 0
              },
              "teamLocation": "Team",
              "teamNickname": "Schlachter"
            },
            "playerIDs": [],
            "home": true
          },
          {
            "score": 0,
            "teamId": 10,
            "team": {
              "division": { "size": 13, "divisionName": "East", "divisionId": 0 },
              "teamAbbrev": "RUSS",
              "waiverRank": 1,
              "teamId": 10,
              "record": {
                "pointsAgainst": 620.96,
                "divisionTies": 0,
                "streakLength": 3,
                "homeTies": 0,
                "overallWins": 1,
                "pointsFor": 440.24,
                "divisionPercentage": 0.16667,
                "homeLosses": 3,
                "awayWins": 1,
                "overallPercentage": 0.16667,
                "overallLosses": 5,
                "divisionLosses": 5,
                "divisionWins": 1,
                "homeWins": 0,
                "awayPercentage": 0,
                "streakType": 2,
                "awayLosses": 2,
                "overallTies": 0,
                "divisionStanding": 13,
                "overallStanding": 13,
                "homePercentage": 0,
                "awayTies": 0
              },
              "teamLocation": "St. Paul",
              "teamNickname": "Russell",
              "logoType": "vector",
              "logoUrl": "http://g.espncdn.com/s/ffllm/logos/HelmetAlphabet-ESPN/Helmet-R.svg"
            },
            "playerIDs": [],
            "home": false
          }
        ],
        "bye": false
      },
      {
        "winner": "undecided",
        "teams": [
          {
            "score": 0,
            "teamId": 3,
            "team": {
              "division": { "size": 13, "divisionName": "East", "divisionId": 0 },
              "teamAbbrev": "CM",
              "waiverRank": 4,
              "teamId": 3,
              "record": {
                "pointsAgainst": 668.9,
                "divisionTies": 0,
                "streakLength": 1,
                "homeTies": 0,
                "overallWins": 2,
                "pointsFor": 631.32,
                "divisionPercentage": 0.33333,
                "homeLosses": 4,
                "awayWins": 2,
                "overallPercentage": 0.33333,
                "overallLosses": 4,
                "divisionLosses": 4,
                "divisionWins": 2,
                "homeWins": 0,
                "awayPercentage": 0,
                "streakType": 2,
                "awayLosses": 0,
                "overallTies": 0,
                "divisionStanding": 10,
                "overallStanding": 10,
                "homePercentage": 0,
                "awayTies": 0
              },
              "teamLocation": "Team",
              "teamNickname": "Code Monkeys",
              "logoType": "customValid",
              "logoUrl": "http://i.imgur.com/DZJH1AY.jpg"
            },
            "playerIDs": [],
            "home": true
          },
          {
            "score": 0,
            "teamId": 12,
            "team": {
              "division": { "size": 13, "divisionName": "East", "divisionId": 0 },
              "teamAbbrev": "KUHL",
              "waiverRank": 6,
              "teamId": 12,
              "record": {
                "pointsAgainst": 622.6,
                "divisionTies": 0,
                "streakLength": 1,
                "homeTies": 0,
                "overallWins": 2,
                "pointsFor": 645.22,
                "divisionPercentage": 0.4,
                "homeLosses": 1,
                "awayWins": 1,
                "overallPercentage": 0.4,
                "overallLosses": 3,
                "divisionLosses": 3,
                "divisionWins": 2,
                "homeWins": 1,
                "awayPercentage": 0,
                "streakType": 2,
                "awayLosses": 2,
                "overallTies": 0,
                "divisionStanding": 8,
                "overallStanding": 8,
                "homePercentage": 0,
                "awayTies": 0
              },
              "teamLocation": "Team",
              "teamNickname": "Kuhl"
            },
            "playerIDs": [],
            "home": false
          }
        ],
        "bye": false
      },
      {
        "winner": "undecided",
        "teams": [
          {
            "score": 0,
            "teamId": 13,
            "team": {
              "division": { "size": 13, "divisionName": "East", "divisionId": 0 },
              "teamAbbrev": "BIRD",
              "waiverRank": 10,
              "teamId": 13,
              "record": {
                "pointsAgainst": 605.58,
                "divisionTies": 0,
                "streakLength": 2,
                "homeTies": 0,
                "overallWins": 4,
                "pointsFor": 666.94,
                "divisionPercentage": 0.66667,
                "homeLosses": 1,
                "awayWins": 2,
                "overallPercentage": 0.66667,
                "overallLosses": 2,
                "divisionLosses": 2,
                "divisionWins": 4,
                "homeWins": 2,
                "awayPercentage": 0,
                "streakType": 2,
                "awayLosses": 1,
                "overallTies": 0,
                "divisionStanding": 4,
                "overallStanding": 4,
                "homePercentage": 0,
                "awayTies": 0
              },
              "teamLocation": "Boo",
              "teamNickname": "Birds",
              "logoType": "customValid",
              "logoUrl": "https://orig00.deviantart.net/1cc0/f/2013/188/e/c/bird_animation_by_ilovebluemorethanyou-d6cerkp.gif"
            },
            "playerIDs": [],
            "home": true
          },
          {
            "score": 0,
            "teamId": 11,
            "team": {
              "division": { "size": 13, "divisionName": "East", "divisionId": 0 },
              "teamAbbrev": "STRZ",
              "waiverRank": 5,
              "teamId": 11,
              "record": {
                "pointsAgainst": 526.06,
                "divisionTies": 0,
                "streakLength": 1,
                "homeTies": 0,
                "overallWins": 2,
                "pointsFor": 577.54,
                "divisionPercentage": 0.4,
                "homeLosses": 2,
                "awayWins": 2,
                "overallPercentage": 0.4,
                "overallLosses": 3,
                "divisionLosses": 3,
                "divisionWins": 2,
                "homeWins": 0,
                "awayPercentage": 0,
                "streakType": 1,
                "awayLosses": 1,
                "overallTies": 0,
                "divisionStanding": 9,
                "overallStanding": 9,
                "homePercentage": 0,
                "awayTies": 0
              },
              "teamLocation": "Guano",
              "teamNickname": "All-Stars",
              "logoType": "customValid",
              "logoUrl": "https://pbs.twimg.com/profile_images/886611343004106752/XcXp5NkX_400x400.jpg"
            },
            "playerIDs": [],
            "home": false
          }
        ],
        "bye": false
      },
      {
        "winner": "undecided",
        "teams": [
          {
            "score": 0,
            "teamId": 9,
            "team": {
              "division": { "size": 13, "divisionName": "East", "divisionId": 0 },
              "teamAbbrev": "TM9",
              "waiverRank": 7,
              "teamId": 9,
              "record": {
                "pointsAgainst": 645.34,
                "divisionTies": 0,
                "streakLength": 1,
                "homeTies": 0,
                "overallWins": 3,
                "pointsFor": 603.76,
                "divisionPercentage": 0.5,
                "homeLosses": 3,
                "awayWins": 2,
                "overallPercentage": 0.5,
                "overallLosses": 3,
                "divisionLosses": 3,
                "divisionWins": 3,
                "homeWins": 1,
                "awayPercentage": 0,
                "streakType": 2,
                "awayLosses": 0,
                "overallTies": 0,
                "divisionStanding": 7,
                "overallStanding": 7,
                "homePercentage": 0,
                "awayTies": 0
              },
              "teamLocation": "Team",
              "teamNickname": "NUMBER 9",
              "logoType": "vector",
              "logoUrl": "http://g.espncdn.com/s/ffllm/logos/8bitHeros-JoeyEllis/8bit_football-01.svg"
            },
            "playerIDs": [],
            "home": true
          },
          {
            "score": 0,
            "teamId": 7,
            "team": {
              "division": { "size": 13, "divisionName": "East", "divisionId": 0 },
              "teamAbbrev": "X69X",
              "waiverRank": 12,
              "teamId": 7,
              "record": {
                "pointsAgainst": 479.64,
                "divisionTies": 0,
                "streakLength": 4,
                "homeTies": 0,
                "overallWins": 4,
                "pointsFor": 583.3,
                "divisionPercentage": 0.8,
                "homeLosses": 0,
                "awayWins": 2,
                "overallPercentage": 0.8,
                "overallLosses": 1,
                "divisionLosses": 1,
                "divisionWins": 4,
                "homeWins": 2,
                "awayPercentage": 0,
                "streakType": 1,
                "awayLosses": 1,
                "overallTies": 0,
                "divisionStanding": 2,
                "overallStanding": 2,
                "homePercentage": 0,
                "awayTies": 0
              },
              "teamLocation": "Elite",
              "teamNickname": "Brown Team",
              "logoType": "customValid",
              "logoUrl": "https://cdn.shopify.com/s/files/1/0879/0890/products/poop1_grande.jpg?v=1486105119"
            },
            "playerIDs": [],
            "home": false
          }
        ],
        "bye": false
      },
      {
        "winner": "undecided",
        "teams": [
          {
            "score": 0,
            "teamId": 2,
            "team": {
              "division": { "size": 13, "divisionName": "East", "divisionId": 0 },
              "teamAbbrev": "HAYN",
              "waiverRank": 2,
              "teamId": 2,
              "record": {
                "pointsAgainst": 733.28,
                "divisionTies": 0,
                "streakLength": 1,
                "homeTies": 0,
                "overallWins": 1,
                "pointsFor": 686.22,
                "divisionPercentage": 0.16667,
                "homeLosses": 3,
                "awayWins": 1,
                "overallPercentage": 0.16667,
                "overallLosses": 5,
                "divisionLosses": 5,
                "divisionWins": 1,
                "homeWins": 0,
                "awayPercentage": 0,
                "streakType": 1,
                "awayLosses": 2,
                "overallTies": 0,
                "divisionStanding": 12,
                "overallStanding": 12,
                "homePercentage": 0,
                "awayTies": 0
              },
              "teamLocation": "Team",
              "teamNickname": "Haynssen",
              "logoType": "vector",
              "logoUrl": "http://g.espncdn.com/s/ffllm/logos/AtTheStadium-RobbHarskamp/At_The_Stadium_01.svg"
            },
            "playerIDs": [],
            "home": true
          },
          {
            "score": 0,
            "teamId": 1,
            "team": {
              "division": { "size": 13, "divisionName": "East", "divisionId": 0 },
              "teamAbbrev": "TIM",
              "waiverRank": 11,
              "teamId": 1,
              "record": {
                "pointsAgainst": 521.12,
                "divisionTies": 0,
                "streakLength": 3,
                "homeTies": 0,
                "overallWins": 4,
                "pointsFor": 569.6,
                "divisionPercentage": 0.8,
                "homeLosses": 1,
                "awayWins": 3,
                "overallPercentage": 0.8,
                "overallLosses": 1,
                "divisionLosses": 1,
                "divisionWins": 4,
                "homeWins": 1,
                "awayPercentage": 0,
                "streakType": 1,
                "awayLosses": 0,
                "overallTies": 0,
                "divisionStanding": 3,
                "overallStanding": 3,
                "homePercentage": 0,
                "awayTies": 0
              },
              "teamLocation": "The Grail",
              "teamNickname": "Marys",
              "logoType": "customValid",
              "logoUrl": "http://cdn-static.denofgeek.com/sites/denofgeek/files/2/26//mrburns.jpeg"
            },
            "playerIDs": [],
            "home": false
          }
        ],
        "bye": false
      },
      {
        "winner": "undecided",
        "teams": [
          {
            "score": 0,
            "teamId": 5,
            "team": {
              "division": { "size": 13, "divisionName": "East", "divisionId": 0 },
              "teamAbbrev": "PART",
              "waiverRank": 13,
              "teamId": 5,
              "record": {
                "pointsAgainst": 499.98,
                "divisionTies": 0,
                "streakLength": 1,
                "homeTies": 0,
                "overallWins": 4,
                "pointsFor": 586.56,
                "divisionPercentage": 0.8,
                "homeLosses": 1,
                "awayWins": 3,
                "overallPercentage": 0.8,
                "overallLosses": 1,
                "divisionLosses": 1,
                "divisionWins": 4,
                "homeWins": 1,
                "awayPercentage": 0,
                "streakType": 1,
                "awayLosses": 0,
                "overallTies": 0,
                "divisionStanding": 1,
                "overallStanding": 1,
                "homePercentage": 0,
                "awayTies": 0
              },
              "teamLocation": "Team",
              "teamNickname": "Partusch"
            },
            "playerIDs": [],
            "home": true
          },
          {
            "score": 0,
            "teamId": 8,
            "team": {
              "division": { "size": 13, "divisionName": "East", "divisionId": 0 },
              "teamAbbrev": "RMRD",
              "waiverRank": 8,
              "teamId": 8,
              "record": {
                "pointsAgainst": 613.12,
                "divisionTies": 0,
                "streakLength": 2,
                "homeTies": 0,
                "overallWins": 3,
                "pointsFor": 592.14,
                "divisionPercentage": 0.6,
                "homeLosses": 1,
                "awayWins": 2,
                "overallPercentage": 0.6,
                "overallLosses": 2,
                "divisionLosses": 2,
                "divisionWins": 3,
                "homeWins": 1,
                "awayPercentage": 0,
                "streakType": 1,
                "awayLosses": 1,
                "overallTies": 0,
                "divisionStanding": 6,
                "overallStanding": 6,
                "homePercentage": 0,
                "awayTies": 0
              },
              "teamLocation": "Team",
              "teamNickname": "Ramrod"
            },
            "playerIDs": [],
            "home": false
          }
        ],
        "bye": false
      }
    ],
    "matchupPeriodId": 8,
    "scoringPeriodId": 8,
    "dateFirstProGameOfScoringPeriod": "2018-10-26T00:20:00.000Z",
    "proGamesInProgress": false,
    "nextProGameId": 381025034,
    "dateNextProGame": "2018-10-26T00:20:00.000Z"
  }
}
