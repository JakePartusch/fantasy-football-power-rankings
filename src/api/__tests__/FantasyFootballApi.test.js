import { FantasyFootballApi } from '../FantasyFootballApi';
import { scoreboard_response } from '../__mock__/espn_api';

describe('FantasyFootballApi Tests', () => {
  describe('ESPN API', () => {
    const api = new FantasyFootballApi();

    it('parse week 1 scoreboard', () => {
      const data = api.buildWeekScores(scoreboard_response(1).scoreboard.matchups);
      expect(data.length).toEqual(13);
      expect(data).toMatchSnapshot();
    });

    it('parse week scoreboard (not played yet - all undecided', () => {
      const data = api.buildWeekScores(scoreboard_response(2).scoreboard.matchups);

      expect(data).toMatchSnapshot();
      expect(data.length).toEqual(0);
    });
  });
});
