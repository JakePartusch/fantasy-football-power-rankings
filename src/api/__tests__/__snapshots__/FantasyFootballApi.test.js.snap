// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FantasyFootballApi Tests ESPN API parse week 1 scoreboard 1`] = `
Array [
  Object {
    "id": 5,
    "name": "Team Partusch",
    "outcome": "BYE",
    "score": 111.54,
  },
  Object {
    "id": 2,
    "name": "Team Haynssen",
    "outcome": "LOSE",
    "score": 125.52,
  },
  Object {
    "id": 8,
    "name": "Team Ramrod",
    "outcome": "WIN",
    "score": 127.04,
  },
  Object {
    "id": 9,
    "name": "Team NUMBER 9",
    "outcome": "LOSE",
    "score": 84.6,
  },
  Object {
    "id": 1,
    "name": "The Grail Marys",
    "outcome": "WIN",
    "score": 125.66,
  },
  Object {
    "id": 13,
    "name": "Boo Birds",
    "outcome": "WIN",
    "score": 106.76,
  },
  Object {
    "id": 7,
    "name": "Elite Brown Team",
    "outcome": "LOSE",
    "score": 69.24,
  },
  Object {
    "id": 3,
    "name": "Team Code Monkeys",
    "outcome": "LOSE",
    "score": 101.4,
  },
  Object {
    "id": 11,
    "name": "Guano All-Stars",
    "outcome": "WIN",
    "score": 124.12,
  },
  Object {
    "id": 4,
    "name": "Team Schlachter",
    "outcome": "WIN",
    "score": 121.42,
  },
  Object {
    "id": 12,
    "name": "Team Kuhl",
    "outcome": "LOSE",
    "score": 101.16,
  },
  Object {
    "id": 6,
    "name": "Team Juberian",
    "outcome": "WIN",
    "score": 85.68,
  },
  Object {
    "id": 10,
    "name": "St. Paul Russell",
    "outcome": "LOSE",
    "score": 59.82,
  },
]
`;

exports[`FantasyFootballApi Tests ESPN API parse week scoreboard (not played yet - all undecided 1`] = `Array []`;
